File: src/lib/scala/Slate.Test/src/test/scala/slate/test/ResultTests.scala
package slate.test

import org.scalatest.{BeforeAndAfter, BeforeAndAfterAll, FunSuite}
import slate.common.{SuccessResult, Result}
import slate.common.results.{ResultFuncs, ResultCode, ResultSupportIn}

/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2015 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */


class ResultTests extends FunSuite with BeforeAndAfter with BeforeAndAfterAll with ResultSupportIn {

  test("can build result") {
    val res = new SuccessResult[String]("user01", 123, msg = Some("created"), ref = Some(20), tag = Some("code12"))
    assert( res.success == true )
    assert( res.msg == Some("created" ))
    assert( res.code == 123 )
    assert( res.get == "user01")
    assert( res.ref == Some(20))
    assert( res.tag == Some("code12"))
  }


  test("can build success") {
    ensure(success(true), true, ResultCode.SUCCESS, "success", true)
  }


  test("can build success with type") {
    ensure(success[Int](123), true, ResultCode.SUCCESS, "success", 123)
  }


  test("can build failure") {
    ensure(failure(), false, ResultCode.FAILURE, "failure", null)
  }


  test("can build unauthorized") {
    ensure(unAuthorized(), false, ResultCode.UNAUTHORIZED, "unauthorized", null)
  }


  test("can build unexpected") {
    ensure(unexpectedError(), false, ResultCode.UNEXPECTED_ERROR, "unexpected error", null)
  }


  test("can build conflict") {
    ensure(conflict(), false, ResultCode.CONFLICT, "conflict", null)
  }


  test("can build not found") {
    ensure(notFound(), false, ResultCode.NOT_FOUND, "not found", null)
  }


  test("can build okFailure with error") {

    val res = ResultFuncs.okOrFailure( {
      val msg = "fail"
      if(msg == "fail" ) {
        throw new IllegalArgumentException("test")
      }
      msg
    })
    ensure(res, false, ResultCode.FAILURE, "test", null)
  }


  test("can map to another value"){
    val yes = ok(msg = Some("m1"), tag = Some("t1"), ref = Some("GUID_123"))
    val good  = yes.map( v => if(v) 1 else 0 )

    assert( good.success )
    assert( good.code == yes.code )
    assert( good.msg == yes.msg )
    assert( good.tag == yes.tag )
    assert( good.ref == yes.ref )
    assert( good.get == 1 )
  }


  test("can map failure"){
    val fail = no(msg = Some("m1"), tag = Some("t1"), ref = Some("GUID_123"))
    val fail2  = fail.map( v => if(v) 1 else 0 )

    assert( !fail2.success )
    assert( fail2.code == fail.code )
    assert( fail2.msg  == Some("m1"))
    assert( fail2.tag  == Some("t1"))
    assert( fail2.ref  == Some("GUID_123"))
    assert( fail2.isEmpty )
  }


  test("can flat map to another value"){
    val yes = ok(msg = Some("m1"), tag = Some("t1"), ref = Some("GUID_123"))
    val good  = yes.flatMap( v => {
      if(v)
        success(1, msg = Some("f_msg1"), tag = Some("f_tag1"), ref = Some("GUID_XYZ"))
      else
        failure[Int](msg = Some("f_msg2"), tag = Some("f_tag2"))
    } )

    assert( good.success )
    assert( good.code == yes.code )
    assert( good.msg  == Some("f_msg1") )
    assert( good.tag  == Some("f_tag1") )
    assert( good.ref  == Some("GUID_XYZ") )
  }


  private def ensure(result:Result[Any], success:Boolean, code:Int, msg:String, data:Any):Unit = {
    assert(result.success == success)
    assert(result.code == code)
    assert(result.msg == Some(msg))
    if(result.success){
      assert(result.get == data)
    }
  }
}

File: src/lib/kotlin/slatekit-tests/src/test/kotlin/test/common/ResultTests.kt
package test.common

import org.junit.Assert
import slatekit.common.results.ResultFuncs.success
import slatekit.common.results.ResultFuncs.failure
import slatekit.common.results.ResultFuncs.unAuthorized
import slatekit.common.results.ResultFuncs.unexpectedError
import slatekit.common.results.ResultFuncs.conflict
import slatekit.common.results.ResultFuncs.notFound
import org.junit.Test
import slatekit.common.*
import slatekit.common.Success
import slatekit.common.results.*
import slatekit.common.results.ResultCode.CONFLICT
import slatekit.common.results.ResultCode.FAILURE
import slatekit.common.results.ResultCode.NOT_FOUND
import slatekit.common.results.ResultCode.SUCCESS
import slatekit.common.results.ResultCode.UNAUTHORIZED
import slatekit.common.results.ResultCode.UNEXPECTED_ERROR

/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2015 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */


//object MyCodes {
//
//  val UserCreated = slatekit.common.results.Success(200, "Created user")
//  val UserUpdated = slatekit.common.results.Success(201, "Updated user")
//  val UserDeleted = slatekit.common.results.Success(200, "Deleted user")
//}


class ResultTests {


  data class Person(val name:String)


  @Test fun can_build_result() {
    val res =  Success("user01", 123, msg = "created")
    assert( res.success  )
    assert( res.msg == "created" )
    assert( res.code == 123 )
    assert( res.getOrElse { "" } == "user01")
  }


  @Test fun can_build_success() {
    ensure(success(true), true, SUCCESS, "success", true)
  }


  @Test fun can_build_success_with_type() {
    ensure(success(123), true, SUCCESS, "success", 123)
  }


  @Test fun can_build_failure() {
    ensure(failure(), false, FAILURE, "failure", null)
  }


  @Test fun can_build_unauthorized() {
    ensure(unAuthorized(), false, UNAUTHORIZED, "unauthorized", null)
  }


  @Test fun can_build_unauthorized2() {
    val result1 = unAuthorized<Int>("1")
    val result2 = unAuthorized<Int, Err>(Invalid(2, "Unauthorized"))

//    val result4 = success2<Person, Err>(Person("kishore"))
//    val result5 = success2<Person, Err>(Person("kishore"), MyCodes.UserCreated)
//    val result6 = failure2<Person, Err>(Invalid(300, "Invalid user"))
  }


  @Test fun can_build_unexpected() {
      val result = unexpectedError<Person>(Exception("Test error"))
      assert(!result.success)
      assert(result.code == UNEXPECTED_ERROR)
      assert(result.msg == "unexpected error")
  }


  @Test fun can_build_conflict() {
    ensure(conflict(), false, CONFLICT, "conflict", null)
  }


  @Test fun can_build_not_found() {
    ensure(notFound(), false, NOT_FOUND, "not found", null)
  }


  @Test
  fun test_success(){
    val res = success(Person("peter"), msg="created")
    Assert.assertTrue(res.success)
    Assert.assertEquals("created", res.msg)
    Assert.assertEquals(SUCCESS, res.code)
  }


  @Test
  fun test_map_success(){
    val res1 = success(Person("peter"), msg="created")
    val res = res1.map { it -> Pair(it, "spider-man") }

    Assert.assertTrue(res.success)
    Assert.assertEquals("created", res.msg)
    Assert.assertEquals(SUCCESS, res.code)

    val user = res.getOrElse( { Pair(Person("kishore"), "reddy") })
    Assert.assertTrue(user.first.name == "peter")
    Assert.assertTrue(user.second == "spider-man")
  }


  @Test
  fun test_map_failure(){
    val res1 = failure<Person>()
    val res = res1.map { it -> Pair(it, "spider-man") }

    Assert.assertFalse(res.success)
    Assert.assertEquals("failure", res.msg)
    Assert.assertEquals(FAILURE, res.code)

    val user = res.getOrElse( { Pair(Person("kishore"), "reddy") })
    Assert.assertTrue(user.first.name == "kishore")
    Assert.assertTrue(user.second == "reddy")
  }


  @Test
  fun test_flat_map_success(){
    val res1 = success(Person("peter"), msg="created")
    val res = res1.flatMap { it -> success(Pair(it, "spider-man"), SUCCESS, "alias") }

    Assert.assertTrue(res.success)
    Assert.assertEquals("alias", res.msg)
    Assert.assertEquals(SUCCESS, res.code)

    val user = res.getOrElse( { Pair(Person("kishore"), "reddy") })
    Assert.assertTrue(user.first.name == "peter")
    Assert.assertTrue(user.second == "spider-man")
  }


  @Test
  fun test_flat_map_failure(){
    val res1 = failure<Person>("setup")
    val res = res1.flatMap { it -> failure<Pair<Person, String>>() }

    Assert.assertFalse(res.success)
    Assert.assertEquals("setup", res.msg)
    Assert.assertEquals(FAILURE, res.code)

    val user = res.getOrElse( { Pair(Person("kishore"), "reddy") })
    Assert.assertTrue(user.first.name == "kishore")
    Assert.assertTrue(user.second == "reddy")
  }


  @Test
  fun test_of_success(){
    val res = Result.of { Person("peter") }

    Assert.assertTrue(res.success)
    Assert.assertEquals("success", res.msg)
    Assert.assertEquals(SUCCESS, res.code)
  }


  @Test
  fun test_of_failure(){
    val res = Result.of { throw Exception("Test error") }

    Assert.assertFalse(res.success)
    Assert.assertEquals("Test error", res.msg)
    Assert.assertEquals(FAILURE, res.code)
  }


  @Test
  fun test_attempt_failure(){
    val res = Result.attempt { throw Exception("Test error") }

    Assert.assertFalse(res.success)
    Assert.assertEquals("Test error", res.msg)
    Assert.assertEquals(UNEXPECTED_ERROR, res.code)
  }


  @Test
  fun test_fold(){
    val res = Result.of { Person("peter") }
    val name = res.fold({ it -> "Success Peter" },
            { it -> "Error Peter"   })

    Assert.assertTrue(res.success)
    Assert.assertEquals("success", res.msg)
    Assert.assertEquals(SUCCESS, res.code)
    Assert.assertEquals("Success Peter", name)
  }


  private fun ensure(result:ResultMsg<Any>, success:Boolean, code:Int, msg:String, data:Any?) {
    assert(result.success == success)
    assert(result.code == code)
    assert(result.msg == msg)
    if(result.success){
      assert(result.getOrElse { "" } == data)
    }
  }
}
